packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -367,31 +371,54 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
         );
     }
 
+    /**
+     * Set up a mapping in the redirect table from fromId to toId. Also, notify the runtime that a reference is added
+     * which is required for GC.
+     */
+    private setRedirection(fromId: string, toId: string | undefined) {
+        this.redirectTable.set(fromId, toId);
+        // Notify runtime of a reference added if toId is not undefined. It can be undefined when a blob is uploaded in
+        // detached mode. In this case, the entry will be updated when the blob is updated.
+        if (toId !== undefined) {
+            this.addedBlobReference(this.getBlobGCNodePath(fromId), this.getBlobGCNodePath(toId));
+        }
+    }
+
     private deleteAndEmitsIfEmpty(id: string) {
-        this.pendingBlobs.delete(id);
-        if (!this.hasPendingBlobs) {
-            this.emit("noPendingBlobs");
+        if (this.pendingBlobs.has(id)) {
+            this.pendingBlobs.delete(id);
+            if (!this.hasPendingBlobs) {
+                this.emit("noPendingBlobs");
+            }
         }
     }
 
     private onUploadResolve(localId: string, response: ICreateBlobResponse) {
+        assert(localId !== undefined, "local ID not available");
         const entry = this.pendingBlobs.get(localId);
         assert(entry?.status === PendingBlobStatus.OnlinePendingUpload ||
             entry?.status === PendingBlobStatus.OfflinePendingUpload,
             0x386 /* Must have pending blob entry for uploaded blob */);
         entry.storageId = response.id;
         if (this.runtime.connected) {
             if (entry.status === PendingBlobStatus.OnlinePendingUpload) {
+                // Send a blob attach op. This serves two purposes:
+                // 1. If its a new blob, i.e., it isn't de-duped, the server will keep the blob alive if it sees this op
+                //    until its storage ID is added to the next summary.
+                // 2. It will create a local ID to storage ID mapping in all clients which is needed to retrieve the
+                //    blob from the server via the storage ID.
+                this.sendBlobAttachOp(localId, response.id);
```

Comment:
Should we check to see if there was a blob attach op for this localId? We wouldn't want to send two attach ops for the same blob if we don't have to right?

Reply:
localIds are unique. We would not be here for 2 different localIds. The entry.status check should take care of this. We can be in that only for newly created blobs.